### PR TITLE
[SYCLomatic #796] Add test for inline device asm parser

### PR DESCRIPTION
**Pull request by yihwang-nv**
Signed-off-by: Wang, Yihan [yihan.wang@intel.com](mailto:yihan.wang@intel.com)

### CODE REVIEW

File: features/feature_case/asm/asm.cu [hunk 1]
```diff
@@ -0,0 +1,1410 @@
+// ====------ asm.cu -------------------------------------- *- CUDA -* ---===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//
+// ===--------------------------------------------------------------------===//
+
```
Comment (zhimingwang36) [inline features/feature_case/asm/asm.cu:22]: As this file is quite big, ~1.5K line, please summary the test plan in the header part, and list out the criteria of test pass.

File: features/feature_case/asm/asm.cu [hunk 2]
```diff
@@ -0,0 +1,1410 @@
...
+      : "f"(f32));
+
+  if (u32 != 1) {
+    *ec = 40;
+    return;
+  }
+}
+
+// clang-format off
+__device__ void slow_lop3(uint32_t &R, uint32_t A, uint32_t B, uint32_t C, uint32_t D) {
```
Comment (zhimingwang36) [inline features/feature_case/asm/asm.cu]:
```suggestion
__device__ void reference_lop3_without_asm(uint32_t &R, uint32_t A, uint32_t B, uint32_t C, uint32_t D) {
```

File: features/feature_case/asm/asm.cu [hunk 3]
```diff
@@ -0,0 +1,1410 @@
...
+  case 251: R = (~A & ~B & ~C) | (~A & ~B & C) | (~A & B & C) | (A & ~B & ~C) | (A & ~B & C) | (A & B & ~C) | (A & B & C); break;
+  case 252: R = (~A & B & ~C) | (~A & B & C) | (A & ~B & ~C) | (A & ~B & C) | (A & B & ~C) | (A & B & C); break;
+  case 253: R = (~A & ~B & ~C) | (~A & B & ~C) | (~A & B & C) | (A & ~B & ~C) | (A & ~B & C) | (A & B & ~C) | (A & B & C); break;
+  case 254: R = A | B | C; break;
+  case 255: R = uint32_t(-1); break;
+  default: break;
+  }
+}
+
+__device__ void fast_lop3(uint32_t &R, uint32_t A, uint32_t B, uint32_t C, uint32_t D) {
```
Comment (zhimingwang36) [inline features/feature_case/asm/asm.cu]:
```suggestion
__device__ void test_lop3(uint32_t &R, uint32_t A, uint32_t B, uint32_t C, uint32_t D) {
```

**Review by wenhuiNi — APPROVED** [review]

File: features/feature_case/asm/asm.cu [hunk 4]
```diff
@@ -0,0 +1,1422 @@
...
+    }
+  }
+  *ec = 0;
+}
+
+int main() {
+  int ret = 0;
+  int *ec = nullptr;
+  cudaMalloc(&ec, sizeof(int));
+
```
Comment (tangjj11) [inline features/feature_case/asm/asm.cu:1391]: not format out of format-off scope.

**Review by tangjj11 — APPROVED** [review]

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**